src/pymanopt/manifolds/generalized_stiefel.py

```diff
@@ -0,0 +1,148 @@
+    Points on the manifold are represented as arrays of shape ``(n, p)``.
+
+    Args:
+        n: The number of rows.
+        p: The number of columns.
+        B: A symmetric positive definite matrix
+        Binv: Inverse of B if known
+        retraction: The type of retraction to use.
+            Possible choices are ``qr`` and ``polar``.
+
+    Note:
+        The matrix :math:`\vmB` can be provided as a numpy array, a
+        scipy sparse matrix, or a scipy linear operator.
+
+        The default retraction used here is a first-order one based on
+        the QR decomposition.
+        To switch to a second-order polar retraction, use
+        ``GeneralizedStiefel(n, p, B, retraction="polar")``.
+
+        Obtaining the Riemannian gradient from the Euclidean gradient requires
+        the inverse of :math:`\vmB`. If this is known, the inverse can be provided
+        to speed up this operation. Otherwise, the LU transorm is computed and
+        stored the first time this function is called. If :math:`\vmB` is given
+        as a scipy linear operator, :math:`\vmB^{-1}` must be supplied.
+    """
+
+    def __init__(
+        self,
+        n: int,
+        p: int,
+        B: np.array,
+        *,
+        Binv: np.array = None,
+        retraction: str = "qr",
+    ):
+        self._n = n
+        self._p = p
+        self.B = B
+        self.Binv = Binv
+        self.LU = None
+        # Check that n is greater than or equal to p
+        if n < p or p < 1:
+            raise ValueError(
+                f"Need n >= p >= 1. Values supplied were n = {n} and p = {p}"
+            )
+        name = f"Generalized Stiefel manifold St({n},{p},B)"
+        dimension = int((n * p - p * (p + 1) / 2))
+        super().__init__(name, dimension)
+        try:
+            self._retraction = getattr(self, f"_retraction_{retraction}")
+        except AttributeError:
+            raise ValueError(f"Invalid retraction type '{retraction}'")
+
+    @property
+    def typical_dist(self):
+        return np.sqrt(self._p)
+
+    def inner_product(self, point, tangent_vector_a, tangent_vector_b):
+        return np.vdot(tangent_vector_a, self.B @ tangent_vector_b)
+
+    def projection(self, point, vector):
+        return vector - point @ self.sym(point.T @ self.B @ vector)
+
+    def sym(self, A):
+        """Returns the symmetric part of A."""
+        return 0.5 * (A + A.T)
+
+    to_tangent_space = projection
+
+    def retraction(self, point, tangent_vector):
+        return self._retraction(point, tangent_vector)
+
+    def _retraction_qr(self, point, tangent_vector):
+        Y = point + tangent_vector
+        return self.gqf(Y)
+
+    def _retraction_polar(self, point, tangent_vector):
+        Y = point + tangent_vector
+        return self.guf(Y)
+
+    def norm(self, point, tangent_vector):
+        return np.sqrt(
+            self.inner_product(point, tangent_vector, tangent_vector)
+        )
+
+    def random_point(self):
```

Comment:
What happens if none of the two cases is true? Point can be undefined. I will use an else statement